cpp/src/arrow/compute/api_scalar.h

```diff
@@ -277,12 +277,16 @@ class ARROW_EXPORT SetLookupOptions : public FunctionOptions {
 class ARROW_EXPORT StructFieldOptions : public FunctionOptions {
  public:
   explicit StructFieldOptions(std::vector<int> indices);
+  explicit StructFieldOptions(std::initializer_list<int>);
+  explicit StructFieldOptions(FieldRef field_ref,
+                              std::vector<int> indices = std::vector<int>());
   StructFieldOptions();
   static constexpr char const kTypeName[] = "StructFieldOptions";
 
   /// The child indices to extract. For instance, to get the 2nd child
   /// of the 1st child of a struct or union, this would be {0, 1}.
   std::vector<int> indices;
+  FieldRef field_ref;
```

Comment:
You probably don't need to keep the `std::vector<int> indices` since the `FieldRef` is able to replace them?

Reply:
In the JIRA issue, [seems it should be kept](https://issues.apache.org/jira/browse/ARROW-17989?focusedCommentId=17617009&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17617009). I tried `ARROW_DEPRECATED` above the public `indices` member which it didn't like since that's one of the places we declare it in the first place. 


> /home/milesg/Projects/arrow/cpp/src/arrow/compute/api_scalar.h:277:20: error: 'arrow::compute::StructFieldOptions::indices' is deprecated: Deprecated since 11.0.0. Use `field_ref` instead. [-Werror=deprecated-declarations]
  277 | class ARROW_EXPORT StructFieldOptions : public FunctionOptions {
      |                    ^~~~~~~~~~~~~~~~~~
/home/milesg/Projects/arrow/cpp/src/arrow/compute/api_scalar.h:289:20: note: declared here
  289 |   std::vector<int> indices;
      |                    ^~~~~~~


Is it better to add the deprecation on the constructors which use `indices`? I initially didn't think so as users can still subsequently be using `StructFieldOptions::indices` without any deprecation notice, and seems anyway constructing from indices ought to be alright.

Reply:
ping @pitrou :)

Reply:
Sorry @milesgranger . I think we should simply remove this member. But we should keep the corresponding constructor.

Reply:
Btw feel free to ask if you need more guidance/help on this :-)

Reply:
Great thanks! I'll take you up on that offer right away. :wink: 

When removing `indices`, I get a compilation error for missing `operator<<` since `FieldRef` is now used in the `kStructFieldOptionsType`'s `DataMember` in `api_scalar.cc`

<details>
<summary>snippet of error</summary>
<\br>
<code>
/home/milesg/Projects/arrow/cpp/src/arrow/compute/function_internal.h:663:3:   required from 'const arrow::compute::FunctionOptionsType* arrow::compute::internal::GetFunctionOptionsType(const Properties& ...) [with Options = arrow::compute::StructFieldOptions; Properties = {arrow::internal::DataMemberProperty<arrow::compute::StructFieldOptions, arrow::FieldRef>}]'
/home/milesg/Projects/arrow/cpp/src/arrow/compute/api_scalar.cc:362:60:   required from here
/home/milesg/Projects/arrow/cpp/src/arrow/compute/function_internal.h:102:6: error: no match for 'operator<<' (operand types are 'std::stringstream' {aka 'std::__cxx11::basic_stringstream<char>'} and 'const arrow::FieldRef')
  102 |   ss << value;
      |   ~~~^~~~~~~~
</code>
</details>

which is reminiscent of [a similar error I'm having in another PR](https://github.com/apache/arrow/pull/14395#discussion_r1012926565)

I _think_ I know how to go about fixing it, implementing a similar `GenericToString` in `function_internal.h`, but as with that PR, not sure where to put that `operator<<` impl.